polyfill.d.ts

```diff
@@ -0,0 +1,30 @@
+/// <reference no-default-lib="true"/>
+
+interface SymbolConstructor {
+  /**
+   * A method that is used with the safe assignment operator `?=`.
+   * Any object that implements the `Symbol.result` method should return `[error, undefined]` or `[null, data]` tuple.
+   */
+  readonly result: unique symbol
+}
+
+interface Function {
+  [Symbol.result]<T extends (...args: any[]) => any>(
+    this: T,
+    ...args: Parameters<T>
+  ): ReturnType<T> extends Promise<infer R>
+    ? Promise<
+        | readonly [NonNullable<unknown> | Error, undefined]
```

Comment:
The type `Error` is assignable to type `NonNullable<unknown>`, that means that:
`NonNullable<unkown> | Error` is the same as `NonNullable<unknown>`

Reply:
Thank you for the review!
You're technically correct that `Error` is already included in `NonNullable<unknown>`, making the union redundant. However, I intentionally used `NonNullable<unknown>` instead of `{}` to be very explicit about the types involved. Since TypeScript doesn't merge `Error `with `{}`, explicitly including it in union ensures that the type clearly signals the intention to handle Error objects specifically. I think it makes the code's intent unmistakable, improving readability and reducing the potential for misunderstandings, particularly in more complex scenarios.
If we don't need this, final types may look even more simplified:

```ts
interface SymbolConstructor {
  /**
   * A method that is used with the safe assignment operator `?=`.
   * Any object that implements the `Symbol.result` method should return `[error, undefined]` or `[null, data]` tuple.
   */
  readonly result: unique symbol
}

interface Function {
  [Symbol.result]<T extends (...args: any[]) => any>(
    this: T,
    ...args: Parameters<T>
  ): [ReturnType<T>] extends [never]
    ? readonly [{}, never]
    : ReturnType<T> extends Promise<never>
      ? readonly [{}, never]
      : ReturnType<T> extends Promise<infer R>
        ? Promise<readonly [{}, undefined] | readonly [null, Awaited<R>]>
        : readonly [{}, undefined] | readonly [null, ReturnType<T>]
}

interface Promise<T = unknown> {
  [Symbol.result](): Promise<
    readonly [{}, undefined] | readonly [null, Awaited<T>]
  >
}
```

Reply:
@valentinpolitov 

So, it just seems you are aiming to "tag" the first tupple argument as an Error ... not more.
I don't see other effect.

But actually it is confuses in the following case:

```ts
const [error, result] ?= someAction();

if (error instanceof Error) {
  /// do something with error
  return; // or throw
}

/// But `result` here is still `T | undefined`
```

Yes, it is an incorrect check if error returned.
But it could be a real mistake of some number of begginers.

So, `Error | {}` (or the same `Error |  NonNullable<unknown>`) provacates mistakes.